plugins/inputs/dpdk/README.md

```diff
@@ -0,0 +1,184 @@
+
+Providing invalid commands will prevent the plugin from starting. Additional commands allow duplicates, but they
+will be removed during execution so each command will be executed only once during each metric gathering interval. 
+
+### Example: Getting metrics from multiple DPDK instances running on same host
+This configuration allows getting metrics from two separate applications exposing their telemetry interfaces
+via separate sockets. For each plugin instance a unique tag `[inputs.dpdk.tags]` allows distinguishing between them. 
+```toml
+# Instance #1 - L3 Forwarding with Power Management Application
+[[inputs.dpdk]]
+  socket_path = "/var/run/dpdk/rte/l3fwd-power_telemetry.v2"
+  device_types = ["ethdev"]
+  additional_commands = ["/l3fwd-power/stats"]
+
+  [inputs.dpdk.ethdev]
+    exclude_commands = ["/ethdev/link_status"]
+
+  [inputs.dpdk.tags]
+    dpdk_instance = "l3fwd-power"
+
+# Instance #2 - L2 Forwarding with Intel Cache Allocation Technology (CAT) Application
+[[inputs.dpdk]]
+  socket_path = "/var/run/dpdk/rte/l2fwd-cat_telemetry.v2"
+  device_types = ["ethdev"]
+
+[inputs.dpdk.ethdev]
+  exclude_commands = ["/ethdev/link_status"]
+
+  [inputs.dpdk.tags]
+    dpdk_instance = "l2fwd-cat"
+```
+This utilizes Telegraf's standard capability of [adding custom tags](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#input-plugins)
+to input plugin's measurements.
+
+## Metrics
+The DPDK socket accepts `command,params` requests and returns metric data in JSON format. All metrics from DPDK socket
+become flattened using [Telegraf's JSON Flattener](../../parsers/json/README.md) and exposed as fields. 
+If DPDK response contains no information (is empty or is null) then such response will be discarded.
+
+> **NOTE:**  Since DPDK allows registering custom metrics in its telemetry framework the JSON response from DPDK 
+> may contain various sets of metrics. While metrics from `/ethdev/stats` should be most stable, the `/ethdev/xstats`
+> may contain driver-specific metrics (depending on DPDK application configuration). The application-specific commands
+> like `/l3fwd-power/stats` can return their own specific set of metrics.  
+
+## Example output
+The output consists of plugin name (`dpdk`), and a set of tags that identify querying hierarchy:
+```
+dpdk,host=dpdk-host,dpdk_instance=l3fwd-power,command=/ethdev/stats,params=0 [fields] [timestamp]
+```
+
+| Tag | Description |
+|-----|-------------|
+| `host` | hostname of the machine (consult [Telegraf Agent configuration](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#agent) for additional details) |
+| `dpdk_instance` | custom tag from `[inputs.dpdk.tags]` (optional) | 
+| `command` | executed command (without params) |
+| `params` | command parameter, e.g. for `/ethdev/stats` it is the id of NIC as exposed by `/ethdev/list`<br>For DPDK app that uses 2 NICs the metrics will output e.g. `params=0`, `params=1`. |
```

Comment:
Since DPDK allows libraries and applications to [register their own telemetry commands](https://doc.dpdk.org/guides-20.05/prog_guide/telemetry_lib.html#registering-commands) and gives freedom in how to use those params (or not to use them at all), it's not always possible to tell what is the actual nature of the param and translate it into a meaningful name like "nic", "device" or anything else. That's why we went with the approach of keeping it generic.

 

To have any hint on params, the only one place we've encountered is the help message for particular command. When registering a command, one can provide such string that can be later on read using `help` invoked for a command, the example would be:
```
--> /help,/ethdev/xstats
{"/help": {"/ethdev/xstats": "Returns the extended stats for a port. Parameters: int port_id"}}
```
Unfortunately those messags are not in an enforced, structured format that would be common across libraries and applications. If this format was standardized we could parse it and extract some detailed information about the nature of the parameter. But currently using such approach would not be reliable.

Reply:
Since DPDK libraries and applications don't provide a standardized format, it's understandable that DPDK has to provide a generic metric format to telegraf.  I'm ok accepting the generic metric format, but I think this should be an area we to try to improve in the future.  Maybe with improvements to json support in telegraf we will be able to come up with a way to use tags and/or metric names instead of field prefixes.

Reply:
No changes here, as we agreed.